File: py_app/visualization/overlay.py
import numpy as np
import cv2
from typing import Dict, List

from py_app.core.datatypes import Detection, OverlayPrediction, CameraConfig, BevConfig
from py_app.visualization.bev import compute_ground_distance
from py_app.core.risk import estimate_risk, risk_to_color


def draw_predictions(
    frame: np.ndarray,
    detections: List[Detection],
    predictions: Dict[int, OverlayPrediction],
    camera_config: CameraConfig,
    bev_config: BevConfig,
) -> np.ndarray:
    canvas = frame
    h, w = canvas.shape[:2]

    for det in detections:
        x1, y1, x2, y2 = det.bbox_xyxy_norm
        px1 = int(np.clip(x1, 0, 1) * w)
        py1 = int(np.clip(y1, 0, 1) * h)
        px2 = int(np.clip(x2, 0, 1) * w)
        py2 = int(np.clip(y2, 0, 1) * h)

        pred = predictions.get(det.track_id)
        geom = compute_ground_distance(frame, det, camera_config, bev_config)

        dist_text = "dist=?"
        risk_score = 0.0

        if geom is not None:
            _foot_x, _foot_y, x_m, _z_m, dist_m = geom
            dist_text = f"{dist_m:.1f}m"
        else:
            x_m, dist_m = None, None

        if pred is None:
            label = f"no-pred {dist_text}"
            color = (255, 255, 255)
        else:
            risk_score = 0.0 if dist_m is None else estimate_risk(
                cross_prob=pred.cross_prob,
                dist_m=dist_m,
                max_range_m=bev_config.max_range_m,
                x_m=x_m,
            )
            pred.risk_score = risk_score
            label = f"P={pred.cross_prob:.2f} Risk={risk_score:.2f} {dist_text}"
            color = risk_to_color(risk_score)

        cv2.rectangle(canvas, (px1, py1), (px2, py2), color, 2)
        cv2.rectangle(canvas, (px1, max(0, py1 - 24)), (min(w - 1, px1 + 260), py1), color, -1)
        cv2.putText(canvas, label, (px1 + 5, max(14, py1 - 7)), cv2.FONT_HERSHEY_SIMPLEX, 0.65, (0, 0, 0), 1, cv2.LINE_AA)

    return canvas

File: ros2_ws/src/pedestrian_visualizer_cpp/src/overlay.cpp
#include "pedestrian_visualizer_cpp/overlay.hpp"

#include <algorithm>
#include <cstdio>

namespace pedestrian_visualizer_cpp
{
    static cv::Scalar risk_to_color(float risk)
    {
        risk = std::clamp(risk, 0.0f, 1.0f);
        int r = static_cast<int>(255.0f * risk);
        int g = static_cast<int>(255.0f * (1.0f - risk));
        return cv::Scalar(0, g, r);
    }

    cv::Mat draw_overlay(
        const cv::Mat& frame_bgr,
        const std::vector<VizTrack>& tracks,
        const std::unordered_map<int, VizPrediction>& predictions)
    {
        cv::Mat canvas = frame_bgr.clone();
        const int h = canvas.rows;
        const int w = canvas.cols;

        for (const auto& tr : tracks) {
            int px1 = std::clamp(static_cast<int>(tr.x1 * w), 0, w - 1);
            int py1 = std::clamp(static_cast<int>(tr.y1 * h), 0, h - 1);
            int px2 = std::clamp(static_cast<int>(tr.x2 * w), 0, w - 1);
            int py2 = std::clamp(static_cast<int>(tr.y2 * h), 0, h - 1);

            auto it = predictions.find(tr.track_id);

            cv::Scalar color(255, 255, 255);
            char label[256];

            if (it == predictions.end()) {
                std::snprintf(label, sizeof(label), "ID=%d no-pred", tr.track_id);
            }
            else {
                const auto& pred = it->second;
                color = risk_to_color(pred.risk_score);
                if (pred.distance_m.has_value()) {
                    std::snprintf(
                        label,
                        sizeof(label),
                        "ID=%d P=%.2f Risk=%.2f %.1fm",
                        tr.track_id,
                        pred.crossing_prob,
                        pred.risk_score,
                        *pred.distance_m);
                }
                else {
                    std::snprintf(
                        label,
                        sizeof(label),
                        "ID=%d P=%.2f Risk=%.2f",
                        tr.track_id,
                        pred.crossing_prob,
                        pred.risk_score);
                }
            }

            cv::rectangle(canvas, cv::Point(px1, py1), cv::Point(px2, py2), color, 2);
            cv::rectangle(
                canvas,
                cv::Point(px1, std::max(0, py1 - 24)),
                cv::Point(std::min(w - 1, px1 + 260), py1),
                color,
                -1);

            cv::putText(
                canvas,
                label,
                cv::Point(px1 + 5, std::max(14, py1 - 7)),
                cv::FONT_HERSHEY_SIMPLEX,
                0.55,
                cv::Scalar(0, 0, 0),
                1,
                cv::LINE_AA);
        }

        return canvas;
    }
}